staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -991,12 +991,14 @@ func (s *store) prepareKey(key string) (string, error) {
 
 // decode decodes value of bytes into object. It will also set the object resource version to rev.
 // On success, objPtr would be set to the object.
-func decode(codec runtime.Codec, versioner storage.Versioner, value []byte, objPtr runtime.Object, rev int64) error {
+func decode(codec runtime.Codec, resource string, key string, versioner storage.Versioner, value []byte, objPtr runtime.Object, rev int64) error {
 	if _, err := conversion.EnforcePtr(objPtr); err != nil {
 		return fmt.Errorf("unable to convert output object to pointer: %v", err)
 	}
 	_, _, err := codec.Decode(value, nil, objPtr)
 	if err != nil {
+		metrics.RecordDecodeError(resource)
+		klog.V(4).Infof("Decoding %s \"%s\" failed: %v", resource, key, err)
```

Comment:
I still have an open question about whether content in `data` can leak into the error and get logged

Reply:
By flipping a protobuf object, I got the logging like
```
I1208 07:56:17.397972 10 store.go:181] Decoding pods "/pods/default/nginx" failed: illegal base64 data at input byte 3
```
The logging of err looks good.
I'm still not quite sure if error would be polluted (at least not during my reproducing), but it makes sense to me that we could avoid potential issues by not logging err.

Reply:
Also, have tried to call bad object multiple times, the metric could accumulate correctly each time
```
$ kubectl get pods nginx
Error from server: illegal base64 data at input

$ kubectl get --raw /metrics | grep storage_decode_errors
# HELP storage_decode_errors [ALPHA] Number of stored object decode errors split by object type.
# TYPE storage_decode_errors counter
storage_decode_errors{type="pods"} 1
```
```
$ kubectl get pods nginx
Error from server: illegal base64 data at input

$ kubectl get --raw /metrics | grep storage_decode_errors
# HELP storage_decode_errors [ALPHA] Number of stored object decode errors split by object type.
# TYPE storage_decode_errors counter
storage_decode_errors{type="pods"} 2
```